source/ui/dataman/AMScanThumbnailGridView.cpp

```diff
@@ -0,0 +1,422 @@
+
+		default:
+			return QRect();
+		}
+	} else {
+		// Is Thumbnail
+		switch(index.column()) {
+		case 1:
+			return geometryManager_->thumbnailImageRectangle(rowContentRect);
+		case 2:
+			return geometryManager_->thumbnailTitleRectangle(rowContentRect);
+
+		default:
+			return QRect();
+		}
+	}
+
+}
+
+void AMScanThumbnailGridView::scrollTo(const QModelIndex &index, QAbstractItemView::ScrollHint hint)
+{
+	// Note: currently ignores the scroll hint. Places the corresponding scan as close to the top of
+	// the view as is possible
+	Q_UNUSED(hint);
+
+	if(!index.isValid())
+		return;
+
+	// If the passed index refers to a thumbnail (i.e. its parent is valid) then we need to
+	// scrollTo the index of the parent scan
+	if(index.parent().isValid()) {
+		scrollTo(index.parent(), hint);
+		return;
+	}
+
+	// Obtian the geometry of the index
+	int itemCellIndex = index.row();
+	QRect indexCellGeometry = geometryManager_->cellGeometryAt(itemCellIndex, horizontalOffset(), verticalOffset());
+
+	verticalScrollBar()->setValue(indexCellGeometry.y());
+	horizontalScrollBar()->setValue(indexCellGeometry.x());
+}
+
+void AMScanThumbnailGridView::onItemSelected(int itemIndex, QItemSelectionModel::SelectionFlags flags)
+{
+	if(itemIndex < 0)
+		return;
+
+	QModelIndex selectedItemIndex = model()->index(itemIndex, 0, QModelIndex());
+	selectionModel()->select(selectedItemIndex, flags | QItemSelectionModel::Rows);
+}
+
+void AMScanThumbnailGridView::onSelectionExtended(int itemIndex)
+{
+	QModelIndex extendToIndex = model()->index(itemIndex, 0, QModelIndex());
+	if(!extendToIndex.isValid())
+		return;
+
+	// Get the lowest currently selected row index
+	int lowestSelectedRowIndex = 0;
+	int rowCount = model()->rowCount();
+	bool lowestSelectedRowIndexFound = false;
+	while(lowestSelectedRowIndex < rowCount && !lowestSelectedRowIndexFound) {
+		if(selectionModel()->isRowSelected(lowestSelectedRowIndex, QModelIndex())) {
+			lowestSelectedRowIndexFound = true;
+		} else {
+			++lowestSelectedRowIndex;
+		}
+	}
+
+	if(lowestSelectedRowIndexFound)	{
+		QModelIndex lowestSelectedModelIndex =
+				model()->index(lowestSelectedRowIndex, 1, QModelIndex());
+
+		if(lowestSelectedModelIndex.isValid()) {
+
+			QItemSelection newSelectedItems;
+
+			if(extendToIndex.row() < lowestSelectedModelIndex.row())
+				newSelectedItems.select(extendToIndex, lowestSelectedModelIndex);
+			else
+				newSelectedItems.select(lowestSelectedModelIndex, extendToIndex);
+
+			selectionModel()->select(newSelectedItems, QItemSelectionModel::Select | QItemSelectionModel::Rows);
+		}
+	}
+}
+
+void AMScanThumbnailGridView::onItemDoubleClicked(int itemIndex)
+{
+	QModelIndex doubleClickedItemIndex = model()->index(itemIndex, 0, QModelIndex());
+	emit doubleClicked(doubleClickedItemIndex);
+}
+
+void AMScanThumbnailGridView::onDragBegun()
+{
+	QDrag* drag = new QDrag(this);
+	QMimeData* mimeData = model()->mimeData(selectionModel()->selectedIndexes());
```

Comment:
The drag operation takes parentage of the mimeData when drag->setMimeData(mimeData) is called. So long as the drag is constructed with a QWidget then the cleanup happens (according to the documentation).